modules/fedimint-lnv2-client/src/lib.rs

```diff
@@ -0,0 +1,725 @@
+                                    Ok(..) => {
+                                        yield SendState::Refunded;
+                                        return;
+                                    },
+                                    Err(..) => {
+                                        // The gateway may have incorrectly claimed the outgoing contract thereby causing
+                                        // our refund transaction to be rejected. Therefore, we check one last time if 
+                                        // the preimage is available before we enter the failure state.
+                                        if let Some(preimage) = module_api.await_preimage(
+                                            &state.contract.contract_id(),
+                                            0
+                                        ).await {
+                                            if state.contract.verify_preimage(&preimage) {
+                                                yield SendState::Success(preimage);
+                                                return;
+                                            }
+                                        }
+
+                                        yield SendState::Failure;
+                                        return;
+                                    },
+                                }
+                            },
+                            SendSMState::Rejected(..) => {
+                                yield SendState::FundingRejected;
+                                return;
+                            },
+                        }
+                    }
+                }
+            }
+        }))
+    }
+
+    pub async fn receive(
+        &self,
+        gateway_api: SafeUrl,
+        invoice_amount: Amount,
+    ) -> anyhow::Result<(Bolt11Invoice, OperationId)> {
+        self.receive_internal(
+            gateway_api,
+            invoice_amount,
+            INVOICE_EXPIRATION_SECONDS_DEFAULT,
+            String::new(),
+            PaymentFee::default(),
+        )
+        .await
+    }
+
+    pub async fn receive_internal(
+        &self,
+        gateway_api: SafeUrl,
+        invoice_amount: Amount,
+        expiry_time: u32,
+        description: String,
+        max_fee: PaymentFee,
+    ) -> anyhow::Result<(Bolt11Invoice, OperationId)> {
+        let (contract, invoice) = self
+            .create_contract_and_fetch_invoice_internal(
+                self.keypair.public_key(),
+                gateway_api,
+                invoice_amount,
+                expiry_time,
+                description,
+                max_fee,
+            )
+            .await?;
+
+        let operation_id = self
+            .receive_external_contract(contract)
+            .await
+            .expect("The contract has been generated with our public key");
+
+        Ok((invoice, operation_id))
+    }
+
+    pub async fn create_contract_and_fetch_invoice(
```

Comment:
Can this function and the one below be made private or do you anticipate higher layers calling this function?